Don't hardcode any keys

rissson opened the pull request:
<!--- Provide a short summary of your changes in the Title above -->

## Description

Alternative to #53 which will avoid issues like this in the future.

## Related Issue
<!--- If you are suggesting a new feature or change, please create an issue first -->
<!--- Please link to the issue, discourse, or stackoverflow here: -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

rissson committed:
Don't hardcode any keys

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson commented:
I should note that this changes the ordering of the keys in the rendered file.

mrlesmithjr commented:
In `tasks/configure.yml:4` – Blocking: combine merges left → right, so the defaults you add on the right side overwrite any user-supplied values (e.g., a caller’s network.version or renderer). Please swap the merge
  order to netplan_default_configuration | combine(netplan_configuration, recursive=True) so role consumers stay in control of their config.

rissson commented:
> In `tasks/configure.yml:4` – Blocking: combine merges left → right, so the defaults you add on the right side overwrite any user-supplied values (e.g., a caller’s network.version or renderer). Please swap the merge order to netplan_default_configuration | combine(netplan_configuration, recursive=True) so role consumers stay in control of their config.

Good catch! I've changed this as required.

mrlesmithjr reviewed (changes requested):
## Request Changes - Needs Broader Discussion ⚠️

This PR proposes a significant architectural change that warrants community discussion before proceeding.

### What's changing:
- **DELETES** the entire `templates/etc/netplan/config.yaml.j2` template
- Replaces template-based rendering with data merging using `combine()`
- Changes fundamental architecture from explicit to implicit configuration

### Concerns:

**1. Breaking Architectural Change**
This fundamentally changes how the role works. The template-based approach has been working reliably and is easier to understand/debug.

**2. Conflicts with PR #53**
PR #53 adds dummy-devices support to the template you're removing. These PRs are mutually exclusive.

**3. Less Explicit**
The current template makes it clear which netplan keys are supported. The merge approach is more implicit, which can make troubleshooting harder.

**4. Testing Gaps**
This needs comprehensive testing across all device types (ethernets, wifis, bonds, bridges, vlans, tunnels, and now dummy-devices) to ensure no regressions.

### Suggested Path Forward:

1. **Open an issue** to discuss this architectural change with the community first
2. If the community agrees with the new approach:
   - Wait for PR #53 to merge first
   - Rebase this PR
   - Ensure dummy-devices support is maintained
   - Add comprehensive tests for all device types
   - Include migration documentation

### Benefits of your approach:
- ✅ More maintainable (no template conditionals)
- ✅ More flexible (handles any netplan key automatically)
- ✅ Cleaner code

The idea has merit, but it's a major change that needs discussion and thorough validation.

What do you think about opening an issue to discuss this approach first?

rissson commented:
> What do you think about opening an issue to discuss this approach first?

Fully agree, I have a bunch of things going on at $dayjob currently, but I will do so as soon as I have a minute of time.

mrlesmithjr reviewed (changes requested):
## Code Review: REQUEST CHANGES

Thank you for this excellent architectural refactor! The approach of using dynamic YAML merging instead of hardcoded template keys is a significant improvement that future-proofs the role against netplan schema changes. However, there are a few issues to address.

---

### Critical Issue

#### 1. Version Override Behavior May Break

**Location:** `defaults/main.yml:14-21`

```yaml
netplan_default_configuration_raw: |
  network:
    version: 2
```

**Problem:** The old template allowed users to override version via `netplan_configuration['network']['version']|default('2')`. The new approach always sets `version: 2` in defaults. If users explicitly set a different version in their configuration, the `combine()` behavior needs verification.

**Suggested fix:**
```yaml
netplan_default_configuration_raw: |
  network:
    version: {{ netplan_configuration['network']['version']|default('2') }}
  {% if netplan_renderer is not none %}
    renderer: {{ netplan_renderer }}
  {% endif %}
```

---

### Major Issues

#### 2. Missing Structure Validation

**Location:** `tasks/configure.yml:4`

If a user provides a malformed configuration without the `network` wrapper:
```yaml
netplan_configuration:
  ethernets:  # Missing 'network' wrapper
    enp0s3:
      dhcp4: true
```

This would create invalid output with `ethernets` at the wrong level.

**Suggested:** Add an assertion to validate the structure:
```yaml
- name: Validate netplan_configuration structure
  ansible.builtin.assert:
    that:
      - netplan_configuration.network is defined or netplan_configuration == {}
    fail_msg: "netplan_configuration must have 'network' as root key"
  when: netplan_configuration != {}
```

---

#### 3. Undocumented Whitespace Fix

**Location:** `tasks/configure.yml:6`

The change from `netplan_config_file_mode}}` to `netplan_config_file_mode }}` is a good fix but should be mentioned in the PR description for completeness.

---

### Minor Issues

#### 4. Documentation Update Needed

The README should be updated to mention:
- The role now supports ALL netplan keys automatically
- Users no longer need to wait for template updates to use new netplan features

**Suggested addition to README:**
```markdown
## Supported Netplan Keys

This role automatically supports all netplan configuration keys through 
dynamic YAML merging. Any valid netplan configuration can be specified 
in `netplan_configuration` under the `network` key.
```

---

### Positive Aspects

1. Eliminates template maintenance for new netplan keys
2. Uses modern Ansible filter approach (`combine`, `to_nice_yaml`)
3. Cleaner, more maintainable code
4. Proper FQCN usage for the copy module

---

### Testing Recommendations

Before merging, please verify:
1. Deep merge behavior with nested configurations
2. Multiple device types in single configuration
3. Empty configuration handling
4. Existing molecule tests (dummy-devices, wireguard) produce identical output

---

Great work on this refactor! Once the version override and validation issues are addressed, this will be ready to merge.

mrlesmithjr reviewed (changes requested):
## Code Review Feedback

Thank you for this PR! The dynamic YAML merging approach is architecturally superior to explicit templates - it automatically supports all netplan keys without template updates. This is a great improvement for maintainability.

However, there are some issues to address before merge.

### Critical Issues (Blocking)

#### 1. Merge Conflict with Master
PR #53 added `dummy-devices` support to the template this PR deletes. Please rebase on master.

The good news: your dynamic approach will automatically support dummy-devices after rebase (no code changes needed).

#### 2. Incorrect `when` Condition
```yaml
when: netplan_configuration != []
```

`netplan_configuration` defaults to `{}` (dict), not `[]` (list). This condition always evaluates to `true`, causing the task to run even with empty configuration.

**Fix:**
```yaml
when: netplan_configuration.network is defined
```

### Major Issues

#### 3. Missing Structure Validation
If a user provides configuration without the `network` wrapper, invalid output is generated:

```yaml
# User provides this (wrong):
netplan_configuration:
  ethernets:
    enp0s3:
      dhcp4: true
```

**Suggested fix - add validation task:**
```yaml
- name: Validate netplan_configuration structure
  ansible.builtin.assert:
    that:
      - netplan_configuration == {} or netplan_configuration.network is defined
    fail_msg: "netplan_configuration must have 'network' as root key"
  when: netplan_configuration != {}
```

### Minor Issues

4. Update README with new capabilities (supports all netplan keys automatically)
5. Add migration notes for existing users

### What Looks Good

- FQCN usage (`ansible.builtin.copy`)
- Whitespace fix on `mode` parameter
- Maintains backup, permissions, handlers
- `combine(recursive=True)` merge order is correct (verified)
- Future-proof design

### Testing Required

After fixes, please run:
- `molecule test -s wireguard`
- `molecule test -s dummy-devices`

Once these issues are addressed, this is ready to merge. The approach is sound!

rissson commented:
Hello, thanks for the additional review, and sorry for taking so long for this, life got in the way.

> #### 1. Merge Conflict with Master

Rebased and fixed conflicts.

> #### 2. Incorrect `when` Condition
> ```yaml
> when: netplan_configuration.network is defined
> ```

Replaced that condition

> #### 3. Missing Structure Validation

Added the assert task.

>     4. Update README with new capabilities (supports all netplan keys automatically)

Not sure where you'd like me to include that, as there is not capabilities section in the README currently.

>     5. Add migration notes for existing users

There are not migration needed as this change is backwards-compatible.

mrlesmithjr reviewed (changes requested):
## Code Review: REQUEST CHANGES

Thank you for the updates, @rissson! The rebase and addressing previous feedback is appreciated. However, there's a critical bug in the validation task that needs to be fixed before merge.

---

### Critical Issue: Broken Assertion Logic

**Location:** `tasks/configure.yml:3-7`

Your assertion has contradictory logic that will never catch invalid configurations:

```yaml
- name: Validate netplan_configuration structure
  ansible.builtin.assert:
    that:
      - netplan_configuration == {} or netplan_configuration.network is defined
    fail_msg: "netplan_configuration must have 'network' as root key"
  when: netplan_configuration != {}
```

**Problem:**
- The `when` condition skips the task when config is empty
- The `that` condition checks if config equals `{}` (which it never will when task runs) OR has network key
- Result: The first part of the OR (`netplan_configuration == {}`) is always false when the task runs, making it redundant

**Fix:**
```yaml
- name: Validate netplan_configuration structure
  ansible.builtin.assert:
    that:
      - netplan_configuration.network is defined
    fail_msg: "netplan_configuration must have 'network' as root key"
  when: netplan_configuration != {}
```

**Test case:** This invalid config should fail but won't with current code:
```yaml
netplan_configuration:
  ethernets:  # Missing 'network' wrapper
    enp0s3:
      dhcp4: true
```

---

### Major Issue: README Documentation Location

> Not sure where you'd like me to include that, as there is not capabilities section in the README currently.

Add a new section after "Role Variables" and before "Example Playbook":

```markdown
## Supported Netplan Keys

This role dynamically supports **all netplan configuration keys** through YAML merging. 
You do not need to wait for role updates to use new netplan features.

Any valid netplan configuration can be specified in `netplan_configuration` under 
the `network` key. The role automatically merges your configuration with sensible defaults.

### Example with custom keys:

\`\`\`yaml
netplan_configuration:
  network:
    version: 2
    openvswitch:  # Any valid netplan key works
      ports:
        - [patch0-1, patch1-0]
\`\`\`
```

---

### Major Issue: Backwards Compatibility Claim

> There are not migration needed as this change is backwards-compatible.

This isn't entirely accurate. There's a behavioral change around **renderer precedence**:

**Old behavior:** Role variable `netplan_renderer` always took precedence  
**New behavior:** User's `netplan_configuration.network.renderer` takes precedence

Most users won't be affected, but those setting renderer in both places will see different behavior.

**Suggested addition to README:**

```markdown
### Configuration Precedence

The role merges your configuration with defaults. User-provided values in 
`netplan_configuration` take precedence over role-level variables like 
`netplan_renderer`.
```

---

### Testing Required

Please run the molecule tests and confirm they pass:

```bash
molecule test -s dummy-devices
molecule test -s wireguard
```

The PR checklist shows both test items are unchecked.

---

### What Looks Good

- Merge order is correct (defaults | combine user config)
- FQCN usage throughout
- Proper conditionals on tunnel tasks
- Whitespace fix on mode parameter
- Successfully rebased with dummy-devices support

The core architecture is excellent - just needs these fixes before merge!

rissson commented:
Sorry about the whitespace changes in the README, I can remove those if needed.

mrlesmithjr reviewed (approved):
## Code Review: APPROVED

Thank you for addressing all the feedback, @rissson!

All issues from the previous review have been properly resolved:

- ✅ **Assertion logic fixed** - Correctly simplified to check only `netplan_configuration.network is defined`
- ✅ **Documentation added** - "Supported Netplan Keys" section clearly explains the dynamic key support
- ✅ **Precedence documented** - "Configuration Precedence" section addresses the renderer behavior change

The implementation is clean, well-documented, and represents a significant architectural improvement that future-proofs the role against netplan schema changes.

Great work on this contribution!
